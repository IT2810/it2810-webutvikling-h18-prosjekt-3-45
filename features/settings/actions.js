/**
 * Set whether or not the feature for a target amount of
 * daily todos is finished.
 */
export const setGoalEnabled = enabled => ({
  type: 'SET_GOAL_ENABLED',
  enabled,
});

/**
 * Set the target daily amount of todos to finish.
 */
export const setTodoGoal = goal => ({
  type: 'SET_TODO_GOAL',
  goal,
});
