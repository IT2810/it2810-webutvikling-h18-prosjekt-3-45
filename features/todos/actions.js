import uuidv4 from 'uuid/v4';

export const addTodo = (text, description, date) => ({
  type: 'ADD_TODO',
  id: uuidv4(),
  text,
  description,
  date,
});

export const addPedometerTodo = (text, description, date, stepsGoal) => ({
  type: 'ADD_TODO',
  id: uuidv4(),
  creationDate: new Date().toGMTString(),
  isPedometer: true,
  text,
  description,
  date,
  stepsGoal,
});

export const updateTodo = (id, data) => ({
  type: 'UPDATE_TODO',
  id,
  data,
});

export const finishTodo = id => ({
  type: 'FINISH_TODO',
  id,
});

export const deleteTodo = id => ({
  type: 'DELETE_TODO',
  id,
});
