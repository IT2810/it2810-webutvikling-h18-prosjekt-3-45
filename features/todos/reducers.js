const todo = (state = {}, action) => {
  switch (action.type) {
    // Handle adding of todos and the undoing of deletion of todos.
    case 'ADD_TODO':
    case 'READD_TODO':
      // Simply add the specified information into an object, with some
      // additional default information.
      return {
        ...state,
        text: action.text,
        description: action.description,
        date: action.date,
        id: action.id,
        done: false,

        // These apply only when tracking the pedometer
        creationDate: action.creationDate,
        isPedometer: action.isPedometer,
        stepsGoal: action.stepsGoal,
      };

    case 'UPDATE_TODO':
      // Do not update the todo if the ID does not match.
      if (action.id !== state.id) {
        return state;
      }

      return {
        // Merge the old and the new data.
        ...state,
        ...action.data,
      };

    case 'FINISH_TODO':
      // Do not finish this todo if the ID does not match.
      if (action.id !== state.id) {
        return state;
      }

      return {
        // Mark the todo as finished.
        ...state,
        done: true,
        finished: action.finished,
      };

    default:
      return state;
  }
};

const todos = (state = [], action) => {
  switch (action.type) {
    case 'ADD_TODO':
      // Add the new todo to the bottom of the list of todos.
      return [...state, todo(undefined, action)];

    case 'READD_TODO':
      // Remove the todo which is marked as finished from the list.
      const filtered = state.filter(todo => todo.id !== action.id);

      return [
        // Readd the old todo in the index in which it used to be.
        ...filtered.slice(0, action.index + 1),
        todo(undefined, action),
        ...filtered.slice(action.index + 1),
      ];

    case 'UPDATE_TODO':
    case 'FINISH_TODO':
      // For these actions, we simply map all todos and
      // pass them the action. The todos check their own
      // identificators and update accordingly.
      return state.map(entry => todo(entry, action));

    case 'DELETE_TODO':
      // Filter away all todos whichc have a different ID than the one to delete.
      return state.filter(entry => entry.id !== action.id);

    case 'REHYDRATE':
      // When data is loaded from AsyncStorage, we add these to the start of
      // the list. Usually, this should mean the new list is equal to the loaded list,
      // but we merge it with the existing todos in case AsyncStorage is really slow,
      // so that the user is somehow able to create extra todos before loading finishes.
      return [...action.state.todos, ...state];

    default:
      return state;
  }
};

export default todos;
