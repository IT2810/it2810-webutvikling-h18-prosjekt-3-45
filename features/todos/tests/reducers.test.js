import todos from '../reducers';
import * as actions from '../actions';

jest.mock('uuid/v4', () => {
  let counter = 0;

  return () => {
    counter += 1;
    return counter;
  };
});

describe('todos reducer', () => {
  it('should return an empty array initially', () => {
    expect(todos(undefined, {})).toEqual([]);
  });

  it('should add a todo', () => {
    expect(
      todos(
        undefined,
        actions.addTodo(
          'Todo text',
          'Todo description',
          'Thu, 18 Oct 2018 17:42:18 GMT',
        ),
      ),
    ).toMatchSnapshot();
  });

  const state = todos(
    undefined,
    actions.addTodo(
      'Todo text',
      'Todo description',
      'Thu, 18 Oct 2018 17:42:18 GMT',
    ),
  );

  it('should add new todo to the end of the array', () => {
    expect(
      todos(
        state,
        actions.addTodo(
          'Second todo',
          'Second todo description',
          'Thu, 18 Oct 2018 17:45:39 GMT',
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should mark the specified todo as done', () => {
    expect(todos(state, actions.finishTodo(state[0].id))[0].done).toEqual(true);
  });

  it('should remove the given todo from the list', () => {
    expect(todos(state, actions.deleteTodo(state[0].id)).length).toEqual(0);
  });
});
