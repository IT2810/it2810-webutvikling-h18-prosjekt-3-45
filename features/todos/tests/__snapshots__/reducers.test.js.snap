// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducer should add a todo 1`] = `
Array [
  Object {
    "creationDate": undefined,
    "date": "Thu, 18 Oct 2018 17:42:18 GMT",
    "description": "Todo description",
    "done": false,
    "id": 2,
    "isPedometer": undefined,
    "stepsGoal": undefined,
    "text": "Todo text",
  },
]
`;

exports[`todos reducer should add new todo to the end of the array 1`] = `
Array [
  Object {
    "creationDate": undefined,
    "date": "Thu, 18 Oct 2018 17:42:18 GMT",
    "description": "Todo description",
    "done": false,
    "id": 1,
    "isPedometer": undefined,
    "stepsGoal": undefined,
    "text": "Todo text",
  },
  Object {
    "creationDate": undefined,
    "date": "Thu, 18 Oct 2018 17:45:39 GMT",
    "description": "Second todo description",
    "done": false,
    "id": 3,
    "isPedometer": undefined,
    "stepsGoal": undefined,
    "text": "Second todo",
  },
]
`;
